cache/stringcache/chained_grouped_cache.go

```diff
@@ -0,0 +1,80 @@
+package stringcache
+
+import "sort"
+
+type ChainedGroupedCache struct {
+	caches []GroupedStringCache
+}
+
+func NewChainedGroupedCache(caches ...GroupedStringCache) *ChainedGroupedCache {
+	return &ChainedGroupedCache{
+		caches: caches,
+	}
+}
+
+func (c *ChainedGroupedCache) ElementCount(group string) int {
+	sum := 0
+	for _, cache := range c.caches {
+		sum += cache.ElementCount(group)
+	}
+
+	return sum
+}
+
+func (c *ChainedGroupedCache) Contains(searchString string, groups []string) []string {
+	groupMatchedMap := make(map[string]struct{})
+
+	for _, cache := range c.caches {
+		for _, group := range cache.Contains(searchString, groups) {
+			groupMatchedMap[group] = struct{}{}
+		}
```

Comment:
Did I understand it correctly, you want to change the function to 
```go
func (c *ChainedGroupedCache) Contains(searchString string, groups []string, map[string]struct{})
```

and the implementation should put matched groups in this map?

Reply:
Yeah that's the idea. It's a pretty common pattern in C/C++.  
Not sure it would speed this up much, so might not be worth it, just thought I'd bring it up in case!